Complete Grindon stories (#248)
ianfhunter/GNOLL: doc/gh-pages/notation/filters.md

# Conditions & Filters

{: .story }
>**GM**: You are lucky. The potion saved your life, cured all your wounds and gaved you an extra boost of energy. As you get up, you realize that you are in a room filled with bags of golden coins.
>

> >**Grindon**: The heart of the dunjeon!..

Reviewer: ```suggestion
>**Grindon**: The heart of the dungeon!
```

>
>**GM**: Yes. Before anyone appear and catch you, you grab as many bags you can in your arms.
>
>**Grindon**: Finally, all this time training in the woods will serve me well. I surely can hold 8 of those nice bags. Let's roll 8d20.
>
>**GM**: True but hold on. The bags are slippery. And small bags containing 4 coins or less may fall from your arms.
>

> >**Grindon**: Okay, let's roll. Please dice, be kind.

Reviewer: ```suggestion
>**Grindon**: Okay, let's roll. Please dice, be kind.  [8d20>4]
```

>

> >*GNOLL*: [{16, 1, 5, 8, 1, 9, 1, 13}]

Reviewer: ```suggestion
>*GNOLL*: [3]
```

Author: Yes... That's the part that bothered me and I wanted your opinion on this.
As inadvertantly highlighted by this story, what it is the expected output?

First of, why is the output `[3]`? Given your other change, you must mean `[5]`, as there is 5 bags with at least 5 coins.
But would we not be interested by the total number of coins, that is `[51]`.

In both cases, we hide the values of the dice, making it impossible to know if we must re-throw. Arguably, with re-throw, maybe the command should have been `[8d20r#1>4]` with the output being `[55]`.
That is less fun for the story and I can rewrite it to comply to this notation.

I don't know if there are cases where you want to look at the values of the dice. That may be a use-case, but with another notation.
Reading Equation 16 below, it seemed that the output was the list of dice matching the criteria. And `[51]` would be something - not existing - like `[sum(8d20>4)]`, and `[5]` seems to actually be `[8d20>4c]`, as described in [Dice counting](https://www.ianhunter.ie/GNOLL/notation/dice_counting.html#counting-dice).
But this output prevented detecting after the throws what were the dice discarded, making it impossible to re-throw if needed. That may just be that the correct call for re-throw is indeed `[8d20r#1>4]`.

You are the expert and leader on this matter. Tell what you want me to do.

Reviewer: You are right, it should actually be 51 and not 3. I was looking at a bug yesterday that used dice counting and I got mixed up!

I've thought about providing the breakdown of individual dice and I think it is something that should be supported in the future. This is a good example of why it would be useful!  I've filed a quick ticket at #282 for now

I think you're right that the Rerolling (while an interesting example of using a combination of notations) is probably taking away from the story as it is right now.

Author: Ok, understood. I understand the logic and I have an idea to adapt the story.
Keeping the dynamic between GM and Grindor, the GM can "make a mistake" and suggest the roll `[8d20>4]`, resulting in 51.
Grindon reports the error and that he can actually do re-roll and switch to `[8d20r#1>4]`, resulting in `52`. And we end with the same comments from the GM and Grindon.

Now, if you think this is too complex and we should focus on the actual doc instead of the short stories, that's entirely ok, as this is really not the heart of your tool :smile: . Particularly, I don't even know if the syntax `8d20r#1>4` is supported.
I will first make the changes to have a minimal version, then the changes with my suggested story, so we can easily pick one commit or the other.

Reviewer: I think it sounds good. It's worth having some more complicated syntax to show what can be done in the software.

I think the syntax would look like `8d20r==1f>4` (8d20, reroll if one, then filter values for ones over 4)

>

> >**GM**: You escape the dunjeon with...

Reviewer: ```suggestion
>**GM**: You escape the dungeon with 3 bags of gold!
```

>

> >**Grindon The Brave**: Not so fast, dear GM. I see lots of 1s here. It is time for my special class feature. I knew it would come handy.

Reviewer: ```suggestion
>**Grindon The Brave**: Not so fast, dear GM. This is my time to shine. I have a special class feature that lets me reroll dice when it's about losing money. I knew it would come handy! 5d20>4!  

```

Reviewer: Yes, disregard the dice suggestion. I made a mistake as I mentioned in another comment

>

> > *GNOLL*: [{2, 4, 3}]

Reviewer: ```suggestion
> *GNOLL*: [2]
```

Reviewer: Maybe not 2, but should be the sum of all dice (according to current logic)

>

> >**GM**: Lucky you, you did not get 1s in your rerolls.

Reviewer: ```suggestion
>**GM**: Lucky you! You were able to catch two more gold coins!
```

>

> >**Grindon The Brave**: I suppose we can say so...

Reviewer: ```suggestion
>**Grindon The Brave**: I'm rich! I'm rich!
```

# Filtering

> In many games, certain logic is only required if a particular value is rolled. For example, the board game One Deck Dungeon [5] contains a scenario

Reviewer: ```suggestion
In many games, certain logic is only required if a particular value is rolled. For example, the board game One Deck Dungeon  contains a scenario
```
My mistake
